apps/src/javalab/JavalabFileHelper.js

```diff
@@ -18,20 +18,80 @@ export function getDefaultFileContents(filename, viewMode) {
+  The orderOfFiles array will contain the 'fileOrder' of each file as they are assigned
+  their tabKey names (such as 'file-0', 'file-1') which is stored in fileMetadata.
+  For example, if sources contains the following:
+  {
+    'Class1.java': {text: '', tabOrder: 2, isVisible: true},
+    'Class2.java': {text: '', tabOrder: 0, isVisible: true},
+    'Class3.java': {text: '', tabOrder: 1, isVisible: true}
+  },
+  then depending on how Object.keys iterates through the keys on sources, we could have
+  fileMetadata assigned: {'file-0': 'Class2.java', 'file-1': 'Class1.java', 'file-2': 'Class3.java'}.
+  The corresponding orderOfFiles would be [0, 2, 1]
+  and orderedTabKeys would be ['file-0', 'file-2', 'file-1'].
+  */
   let fileMetadata = {};
   let orderedTabKeys = [];
-
+  let unorderedTabKeys = [];
+  let orderOfFiles = [];
+  let fileIndex = 0;
+  let isValid = true;
   Object.keys(sources).forEach(file => {
     if (sources[file].isVisible || isEditingStartSources) {
-      // tabIndex should be the length of orderedTabKeys, not the index
-      // from the list, because we skip hidden files in the tabs.
-      const tabIndex = orderedTabKeys.length;
-      let tabKey = getTabKey(tabIndex);
+      let tabKey = getTabKey(fileIndex);
       fileMetadata[tabKey] = file;
-      orderedTabKeys.push(tabKey);
+      unorderedTabKeys.push(tabKey);
+      let tabOrder = sources[file].tabOrder;
+      // files that are stored may not currently have an tabOrder assigned so that
+      // order is undefined
+      if (Number.isInteger(tabOrder)) {
+        orderOfFiles.push(sources[file].tabOrder);
+      } else {
+        isValid = false;
+      }
+      fileIndex++;
     }
   });
+  const numFiles = fileIndex;
+  // check orderOfFiles for duplicates
+  isValid = isValid && new Set(orderOfFiles).size === orderOfFiles.length;
 
+  /*
+  Assign orderedTabKeys the ordering of files stored in orderOfFiles corresponding to
+  file names stored in fileMetadata.
+  There may be missing tab orders due to validation or support files saved in start_sources
+  mode that are not included outside of start_sources mode.
+  For example, if a levelbuilder created a support file, validation file and 2 starter files,
+  and ordered the file tabs as: 'Validation.java', 'Starter1.java', 'Support.java', 'Starter2.java',
+  then the corresponding tabOrders for each file would be: 0, 1, 2, 3. When isEditingStartSources
+  is false, only 'Start1.java' and 'Starter2.java' would be loaded into sources with
+  the corredponding tabOrders would be 1, 3.
+  If isValid is false due to duplicate tabOrders or any undefined tabOrders, then
+  assign orderedTabKeys the order ['file-0', 'file-1', 'file-2', ...]
+  */
+  if (isValid) {
+    let count = 0;
+    let i = 0;
+    while (count < numFiles) {
+      let index = orderOfFiles.indexOf(i);
+      if (index > -1) {
+        orderedTabKeys.push(unorderedTabKeys[index]);
+        count++;
+      }
+      i++;
```

Comment:
index, count, and i are all often used as aliases for "array index." Give these more descriptive names indicating what they do in each loop.

Reply:
Great suggestion - will do!